inception/inception-api-annotation-experimental/src/main/java/de/tudarmstadt/ukp/inception/experimental/api/AnnotationSystemAPIImpl.java

```diff
@@ -17,402 +17,669 @@
+            throw new Exception();
+        }
+    }
 
-        List<Annotation> annotations = new ArrayList<>();
+    public List<Arc> filterArcs(List<Arc> aRelations, CAS aCas, Viewport aViewport) throws Exception
+    {
+        try {
+            List<Arc> filteredArcs = new ArrayList<>();
+
+            for (Arc arc : aRelations) {
+                AnnotationFS annotation = selectAnnotationByAddr(aCas, arc.getGovernorId().getId());
+
+                for (int i = 0; i < aViewport.getViewport().size(); i++) {
+                    for (int j = aViewport.getViewport().get(i).get(0); j <= aViewport.getViewport()
+                            .get(i).get(1); j++) {
+                        if (annotation.getBegin() == j) {
+                            filteredArcs.add(arc);
+                            break;
+                        }
+                    }
+                }
+            }
+            return filteredArcs;
 
-        for (AnnotationListItem annotationListItem : items) {
-            AnnotationFS annotation = WebAnnoCasUtil.selectAnnotationByAddr(aCas,
-                annotationListItem.getAddr());
-            annotations.add(new Annotation(annotation._id(), annotation.getCoveredText(),
-                annotation.getBegin(), annotation.getEnd(),
-                annotation.getType().getShortName()));
         }
-
-        return annotations;
+        catch (Exception e) {
+            LOG.error("Exception has been thrown during filterArcs()", e);
+            throw new Exception();
+        }
     }
 
-    @Override
-    public List<Annotation> filterAnnotations(List<Annotation> aAnnotations, int[][] aViewport)
+    public String getColorForAnnotation(AnnotationFS aAnnotation, Project aProject)
     {
-        // TODO correct filtering
-        return aAnnotations;
+        // TODO proper coloring strategy (without VTypes)
+        try {
+            TypeAdapter adapter = annotationService.getAdapter(annotationService.findLayer(
+                    projectService.getProject(aProject.getId()), aAnnotation.getType().getName()));
+
+            return PALETTE_NORMAL_FILTERED[adapter.getLayer().getId().intValue()
+                    % PALETTE_NORMAL_FILTERED.length];
+        }
+        catch (Exception e) {
+            LOG.info(
+                    "First entry of PALETTE_NORMAL_FILTERED WAS PICKED DUE TO ERROR. SEE SERVER LOG FOR FURTHER DETAILS.");
+            return PALETTE_NORMAL_FILTERED[0];
+        }
     }
 
-    @Override
-    public void createErrorMessage(String aMessage, String aUser) throws IOException
+    public Pair<List<Span>, List<Arc>> getAnnotations(CAS aCas, long aProjectId,
+            List<AnnotationLayer> aValidLayers)
     {
-        annotationProcessAPI.handleSendErrorMessage(new ErrorMessage(aMessage), aUser);
+        List<Span> spans = new ArrayList<>();
+        List<Arc> relations = new ArrayList<>();
 
+        for (FeatureStructure fs : selectAllFS(aCas)) {
```

Comment:
You have a list of the layers that you want to select the annotations for. You should then get a type adapter for the layer and from the type adapter obtain use `getAnnotationTypeName()` to get the annotation type (cf. de.tudarmstadt.ukp.clarin.webanno.api.annotation.rendering.SpanRenderer.typeSystemInit(TypeSystem)). Then you can iterate over the annotations of each type separately. There is no need to iterate over ALL feature structures - this is very inefficient.